src/main/java/net/mcreator/blockly/java/ProcedureCodeOptimizer.java

```diff
@@ -0,0 +1,95 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package net.mcreator.blockly.java;
+
+@SuppressWarnings("unused") public class ProcedureCodeOptimizer {
+	enum ParseState {
+		INSIDE_INLINE_COMMENT, INSIDE_COMMENT_BLOCK, INSIDE_STRING, INSIDE_STRING_ESCAPE_SEQUENCE, OUTSIDE
+	}
+
+	/**
+	 * This method attempts to remove the parentheses surrounding the given code, if they are paired.
+	 * @param code The code to optimize
+	 * @return If possible, the code without surrounding parentheses
+	 */
+	public static String removeParentheses(String code) {
+		String toClean = code.strip();
+		if (toClean.startsWith("(") && toClean.endsWith(")")) {
+			var state = ParseState.OUTSIDE;
+			int parentheses = 1;
+			char prevChar = '(';
+			int backSlashCounter = 0;
+			for (int i = 1; i < toClean.length() - 1; i++) {
+				char c = toClean.charAt(i);
+				switch (state) {
+				case OUTSIDE:
+					if (c == '/' && prevChar == '/')
+						state = ParseState.INSIDE_INLINE_COMMENT;
+					else if (c == '*' && prevChar == '/')
+						state = ParseState.INSIDE_COMMENT_BLOCK;
+					else if (c == '"')
+						state = ParseState.INSIDE_STRING;
+					else if (c == '(')
+						parentheses++;
+					else if (c == ')' && --parentheses == 0) // The first ( isn't paired with the last ), we can't remove them
+						return code;
+					break;
+				case INSIDE_INLINE_COMMENT:
+					if (c == '\n' || c == '\r')
+						state = ParseState.OUTSIDE;
+					break;
+				case INSIDE_STRING_ESCAPE_SEQUENCE:
+					if (c == '\\') {
+						backSlashCounter++;
+						break;
+					}
+					state = ParseState.INSIDE_STRING;
+				case INSIDE_STRING:
+					if (c == '\\') {
+						state = ParseState.INSIDE_STRING_ESCAPE_SEQUENCE;
+						backSlashCounter = 0;
+					} else if (c == '"' && (prevChar != '\\' || backSlashCounter % 2 != 0)) {
+						state = ParseState.OUTSIDE;
+					}
+					break;
+				case INSIDE_COMMENT_BLOCK:
+					if (c == '/' && prevChar == '*')
+						state = ParseState.OUTSIDE;
+					break;
+				}
+				prevChar = c;
+			}
+			if (state == ParseState.OUTSIDE && --parentheses == 0) // The last character is a valid ")"
+				return toClean.substring(1, toClean.length() - 1);
+		}
+		return code;
+	}
+
+	/**
+	 * This method removes the blockstate/itemstack marker comment, and if possible the parentheses surrounding the
+	 * remaining code.
+	 * @param code The code to optimize
+	 * @return The code without the initial comment and, if possible, without surrounding parentheses
+	 */
+	public static String removeCommentAndParentheses(String code) {
+		String withoutComment = code.replaceFirst("/\\*@BlockState\\*/", "").replaceFirst("/\\*@ItemStack\\*/", "");

```

Comment:
Why is this comment removal needed? Wouldn't it mess with the underlying procedures?

Reply:
How about other cases that don't use this method but use parentheses optimization? Will these work with these marker comments?

Reply:
The other optimized blocks don't accept blockstates/itemstack or call one of the mapping functions in mcitems.ftl (which remove the marker comment), so it isn't an issue

Reply:
I see. In this case we just need to be careful in the future where we use this block and where not :) Or maybe change removeCommentAndParentheses to remove the comment, then process parentheses, and then add it back if it was there before. This way it can be used without this worry too.

Reply:
I don't think re-adding the comment is needed, the user reading the generated code shouldn't have to see itemstack/blockstate markers. I changed the method documentation a little, so that it's clearer when it should be used

Reply:
I meant for other use cases where markers are still needed down the processing line. We are not using this for these cases yet, but may in the future.

Also I would consider adding a global replace function in the end of the procedure code generation pipeline so we cleanup all the markers at the end.